electrumsv/network.py

```diff
@@ -1027,305 +290,144 @@ async def _main_task(self) -> None:
-                if server.protocol == protocol and server.state.can_retry(now)]
 
-    def _random_server_nowait(self, protocol: str) -> Optional[SVServer]:
-        servers = self._available_servers(protocol)
-        return random.choice(servers) if servers else None
+    def get_esv_server_tip_hash(self) -> Optional[bytes]:
+        if not self.esv_client_cached_tip:
+            return None
+        assert self.esv_client_cached_tip is not None
+        return cast(bytes, self.esv_client_cached_tip.hash)
 
-    async def _random_server(self, protocol: str) -> SVServer:
+    # TODO(post-1.4.0) Need a sound system for banning and selecting a better provider
+    async def _initial_headers_sync(self) -> None:
         while True:
-            server = self._random_server_nowait(protocol)
-            if server:
-                return server
-            await sleep(10)
-
-    async def _main_session(self) -> SVSession:
+            try:
+                while self.get_local_height() < await self.get_esv_server_tip_height():
+                    missing_heights = [height for height in range(
+                        self.get_local_height(), await self.get_esv_server_tip_height() + 1)]
+                    await self._fetch_missing_headers_at_heights(missing_heights)
+
+                assert self.esv_client_cached_tip is not None
+                if self.get_local_tip_hash() != self.get_esv_server_tip_hash():
+                    # Wallet can still operate like this so log error and return from function
+                    # TODO(1.4.0) Switch to alternative service providers
+                    logger.error("The Headers API tip is either lagging our previously "
+                        "stored tip or is an equal height fork. Local tip height: %s, "
+                        "Server tip height: %s", self.get_local_height(),
+                        self.esv_client_cached_tip.height)
+                return None
+            except ServiceUnavailableError:
+                logger.info("Headers API service is currently unavailable, retrying in 10 seconds")
+                await asyncio.sleep(10)
+
+    # TODO: periodically check that selected / prioritised server is still the best
+    #  candidate and switch if needed. This ideally should be as simple as polling
+    #  get_esv_headers_client() which handles the selection and prioritisation algorithm.
+    #  For now, we stick to the same server until it disconnects or misbehaves.
+    async def _monitor_chain_tip_task(self) -> None:
+        WAIT_TIME = 10
         while True:
-            session = self.main_session()
-            if session:
-                return session
-            await self.sessions_changed_event.wait()
+            try:
+                # Must always clear cached values when switching to a new server
+                self.esv_client = None
+                self.esv_client_cached_tip = None
+                esv_client = await self.get_esv_headers_client()
```

Comment:
This function needs to be renamed and to be reworked. Having to clear some cached values before calling it, let alone what it even means for it to fail is not good style.